### PR TITLE
Viewi pages support

### PR DESCRIPTION
**Pull request by ivanvoitovych**
Hi @kenjis , please review my draft work. My knowledge of CI4 is limited, so I left you comments related to the framework. I hope this will help to you to finish the integration. 

If there is any questions or my help is needed, I will be glad to provide such.

Looking forward to test the final result :)

### CODE REVIEW

File: app/Adapters/ViewiCodeIgniterAdapter.php [hunk 1]
```diff
@@ -21,18 +21,7 @@ public function __construct(CodeIgniter $app)
...
-        // replace route params `{name}` with placeholders.
-        // @TODO now supports only `{name}`.
-        $url = preg_replace('/{\w+?}/', '(:segment)', $url);
-
-        $routes->{$method}($url, static function (...$params) use ($component) {
-            $controller = new ViewiCodeIgniterComponent($component);
-
-            return $controller->index($params);
-        }, ['as' => $component]);
+        // skip, we use 404 handler to catch all Viewi routes. No need to convert route syntax.
```
Comment (kenjis) [inline app/Adapters/ViewiCodeIgniterAdapter.php]: There are Pros and Cons to this way.
Pros are that there is no need to convert the syntax of the route.
Cons are that CI4 [controller filters](https://codeigniter4.github.io/CodeIgniter4/incoming/filters.html) cannot be used.

**Comment by kenjis** [comment]
@ivanvoitovych Thank you!
I will consider this PR. 
This PR (and Viewi) helps me a lot.

File: app/Adapters/Viewi.php [hunk 2]
```diff
@@ -17,6 +20,20 @@ public static function init(CodeIgniter $app): void
     {
         $adapter = new ViewiCodeIgniterAdapter($app);
         Route::setAdapter($adapter);
+
+        /** @var RouteCollection $routes */
+        $routes = Services::routes();
+        $routes->set404Override(static function () {
+            $request = Services::request();
+            $url = $request->getUri()->getPath();
+            $method = $request->getMethod(); // deprecated, is there any other way to get the method?
```
Comment (kenjis) [inline app/Adapters/Viewi.php]: The parameter is deprecated. The method itself will remain.

File: app/Adapters/Viewi.php [hunk 3]
```diff
@@ -17,6 +20,20 @@ public static function init(CodeIgniter $app): void
...
+            $viewiResponse = Router::handle($url, $method);
+            if (is_string($viewiResponse)) { // html
+                echo $viewiResponse; // TODO: how to return the Response object, not echoing the output. Use case: Viewi gets the HTML page of some component (with HttpClient), needed as a string, not echo.
```
Comment (kenjis) [inline app/Adapters/Viewi.php]: `return $viewiResponse;`

File: app/Config/Routes.php [hunk 4]
```diff
@@ -67,3 +69,6 @@
 if (is_file(APPPATH . 'Config/' . ENVIRONMENT . '/Routes.php')) {
     require APPPATH . 'Config/' . ENVIRONMENT . '/Routes.php';
 }
+
+// Viewi here, after all other routes
+\App\Adapters\Viewi::init(Services::codeigniter()); // is there any better way to get the $app instance ??
```
Comment (kenjis) [inline app/Config/Routes.php:74]: No.

File: app/Adapters/ViewiCodeIgniterAdapter.php (hunk 1 above)
Comment (ivanvoitovych) [reply app/Adapters/ViewiCodeIgniterAdapter.php]: Got your point. Btw, Viewi has middleware/guard support
https://viewi.net/docs/guards
Theoretically, it supposed to cover the filter case. But I will try and make the syntax converter, got an idea.

File: app/Adapters/Viewi.php (hunk 2 above)
Comment (ivanvoitovych) [reply app/Adapters/Viewi.php]: Got it

File: app/Adapters/ViewiCodeIgniterAdapter.php (hunk 1 above)
Comment (ivanvoitovych) [reply app/Adapters/ViewiCodeIgniterAdapter.php]: @kenjis done, please check now

File: app/Adapters/Viewi.php (hunk 3 above)
Comment (ivanvoitovych) [reply app/Adapters/Viewi.php]: done

File: app/Adapters/ViewiCodeIgniterAdapter.php [hunk 5]
```diff
@@ -21,7 +21,57 @@ public function __construct(CodeIgniter $app)
...
+            // collect params
+            $viewiParams = [];
+            foreach ($paramNames as $i => $name) {
+                if ($i < count($params)) {
+                    $viewiParams[$name] = $params[$i];
+                }
+            }
+            return $controller->index($viewiParams);
+        }, ['as' => $component]); 
+        // Question: Has 'as' parameter to be unique ?
```
Comment (kenjis) [inline app/Adapters/ViewiCodeIgniterAdapter.php]: Yes. `as` defines the route name and it must be unique in a HTTP method.

```console
$ php spark routes

CodeIgniter v4.2.7 Command Line Tool - Server Time: 2022-10-16 03:17:02 UTC-05:00

+--------+--------------------+------------------------------------+-----------+----------------+---------------+
| Method | Route              | Name                               | Handler   | Before Filters | After Filters |
+--------+--------------------+------------------------------------+-----------+----------------+---------------+
| GET    | api/posts/([0-9]+) | »                                  | (Closure) |                | toolbar       |
| GET    | /                  | Components\Views\Home\HomePage     | (Closure) |                | toolbar       |
| GET    | counter            | Components\Views\Pages\CounterPage | (Closure) |                | toolbar       |
| GET    | todo               | Components\Views\Pages\TodoAppPage | (Closure) |                | toolbar       |
| GET    | posts/{postId}     | Components\Views\Posts\PostsPage   | (Closure) |                | toolbar       |
+--------+--------------------+------------------------------------+-----------+----------------+---------------+
```

**Comment by kenjis** [comment]
Please run `composer cs-fix` to fix coding style.

**Comment by ivanvoitovych** [comment]
> 

done

**Review by kenjis — APPROVED** [review]
LGTM!

**Comment by kenjis** [comment]
Thank you!

**Comment by kenjis** [comment]
@ivanvoitovych If I navigate to`http://localhost:8080/not-found-page`,
the response status code is 200. How can I change it to 404?

**Comment by ivanvoitovych** [comment]
> @ivanvoitovych If I navigate to`http://localhost:8080/not-found-page`, the response status code is 200. How can I change it to 404?

@kenjis Viewi does not support status code, it should be responsibility of the framework, let me think of something.

**Comment by ivanvoitovych** [comment]
> @ivanvoitovych If I navigate to`http://localhost:8080/not-found-page`, the response status code is 200. How can I change it to 404?

@kenjis 
please check this workaround:
https://github.com/ivanvoitovych/ci4-viewi-demo/commit/02cf81b549c3677e187cd6495f4fa5a2347ebe64

**Comment by kenjis** [comment]
@ivanvoitovych Thank you!
Added https://github.com/kenjis/ci4-viewi-demo/commit/71d1c87b1bb38e012cd0c0ede190d555cbb26aa6

**Comment by ivanvoitovych** [comment]
@kenjis  I will work on adding more elegant way of supporting status codes, perhaps as an optional parameter in Viewi route.

**Comment by ivanvoitovych** [comment]
@kenjis here it is another workaround, using $defaults

https://github.com/ivanvoitovych/ci4-viewi-demo/commit/0d2850506202cbfec229b02a7d46ec2ed59b01e2

looks more elegant, what do you think?

**Comment by kenjis** [comment]
@ivanvoitovych Looks better!
Can  you send a PR?

**Comment by ivanvoitovych** [comment]
> @ivanvoitovych Looks better! Can you send a PR?

sure, here you go https://github.com/kenjis/ci4-viewi-demo/pull/2

**Pull request merged**